### ISSUE TITLE
Do you try to reprodunce 'Deep feature flow for video recognition'?

### ISSUE THREAD

**Issue by TyroneLi** [Scalsol/mega.pytorch#5]
I use your repr and modify some code trying to reproduce DFF. But I only get a little better than 69%.
Here is some main code modified for this part:
generalized_rcnn_dff.py
`
# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
"""
Implements the Generalized R-CNN framework
"""
from PIL import Image
from collections import deque

import torch
from torch import nn
import torch.nn.functional as F

from mega_core.structures.image_list import to_image_list

from ..backbone import build_backbone, build_flownet, build_embednet
from ..rpn.rpn import build_rpn
from ..roi_heads.roi_heads import build_roi_heads

class GeneralizedRCNNDFF(nn.Module):
    """
    Main class for Generalized R-CNN. Currently supports boxes and masks.
    It consists of three main parts:
    - backbone
    - rpn
    - heads: takes the features + the proposals from the RPN and computes
        detections / masks from it.
    """

    def __init__(self, cfg):
        super(GeneralizedRCNNDFF, self).__init__()

        self.backbone = build_backbone(cfg)
        self.flownet = build_flownet(cfg)
        self.rpn = build_rpn(cfg, self.backbone.out_channels)
        self.roi_heads = build_roi_heads(cfg, self.backbone.out_channels)

        self.device = cfg.MODEL.DEVICE
        # 10
        self.all_frame_interval = cfg.MODEL.VID.DFF.ALL_FRAME_INTERVAL
        # 1
        self.key_frame_location = cfg.MODEL.VID.DFF.KEY_FRAME_LOCATION
        self.images = deque(maxlen=self.all_frame_interval)
        self.features = deque(maxlen=self.all_frame_interval)

    def get_grid(self, flow):
        m, n = flow.shape[-2:]
        shifts_x = torch.arange(0, n, 1, dtype=torch.float32, device=flow.device)
        shifts_y = torch.arange(0, m, 1, dtype=torch.float32, device=flow.device)
        shifts_y, shifts_x = torch.meshgrid(shifts_y, shifts_x)

        grid_dst = torch.stack((shifts_x, shifts_y)).unsqueeze(0)
        workspace = torch.tensor([(n - 1) / 2, (m - 1) / 2]).view(1, 2, 1, 1).to(flow.device)

        flow_grid = ((flow + grid_dst) / workspace - 1).permute(0, 2, 3, 1)

        return flow_grid

    def resample(self, feats, flow):
        flow_grid = self.get_grid(flow)
        warped_feats = F.grid_sample(feats, flow_grid, mode="bilinear", padding_mode="border")

        return warped_feats

    def forward(self, images, targets=None):
        """
        Arguments:
            #images (list[Tensor] or ImageList): images to be processed
            targets (list[BoxList]): ground-truth boxes present in the image (optional)

        Returns:
            result (list[BoxList] or dict[Tensor]): the output from the model.
                During training, it returns a dict[Tensor] which contains the losses.
                During testing, it returns list[BoxList] contains additional fields
                like `scores`, `labels` and `mask` (for Mask R-CNN models).

        """
        if self.training and targets is None:
            raise ValueError("In training mode, targets should be passed")

        if self.training:
            images["cur"] = to_image_list(images["cur"])
            images["ref"] = [to_image_list(image) for image in images["ref"]]

            return self._forward_train(images["cur"], images["ref"], targets, images['eq'])
        else:
            images["cur"] = to_image_list(images["cur"])
            # images["ref"] = [to_image_list(image) for image in images["ref"]]

            infos = images.copy()
            infos.pop("cur")
            return self._forward_test(images["cur"], infos)

    def _forward_train(self, img, imgs_ref, targets, eq):
        # 1
        num_refs = len(imgs_ref)
        concat_imgs = torch.cat([img.tensors, *[img_ref.tensors for img_ref in imgs_ref]], dim=0)
        img_cur, imgs_ref = torch.split(concat_imgs, (1, num_refs), dim=0)
        conv_feat = self.backbone(imgs_ref)[0]

        # calculate flow and warp the feature
        concat_imgs_pair = torch.cat([img_cur / 255, imgs_ref / 255], dim=1)

        flow = self.flownet(concat_imgs_pair)
        warped_feats_refs = self.resample(conv_feat, flow)

        if not eq:
            feats = warped_feats_refs
        else:
            feats = conv_feat
        feats = torch.unsqueeze(feats, dim=1)

        proposals, proposal_losses = self.rpn(img, feats, targets)

        if self.roi_heads:
            x, result, detector_losses = self.roi_heads(feats, proposals, targets)
        else:
            detector_losses = {}

        losses = {}
        losses.update(detector_losses)
        losses.update(proposal_losses)
        return losses

    def _forward_test(self, imgs, infos, targets=None):
        """
        forward for the test phase.
        :param imgs:
        :param frame_category: 0 for start, 1 for normal
        :param targets:
        :return:
        """
        def update_feature(img=None, feats=None, embeds=None):
            self.images.append(img)
            # self.features.append(torch.cat([feats, embeds], dim=1))
            self.features.append(feats)

        if targets is not None:
            raise ValueError("In testing mode, targets should be None")

        if infos["frame_category"] == 0:  # a new video
            self.seg_len = infos["seg_len"]
            self.end_id = 0

            self.images = deque(maxlen=self.all_frame_interval)
            self.features = deque(maxlen=1)

            feats_cur = self.backbone(imgs.tensors)[0]
            while len(self.images) < self.key_frame_location + 1:
                update_feature(imgs.tensors, feats_cur)
            
            final_feat = feats_cur

        elif infos["frame_category"] == 1: # not key frame
            self.end_id = min(self.end_id + 1, self.seg_len - 1)
            # end_image = infos["ref"][0].tensors

            # update_feature(end_image)

            ref_img = self.images[0]
            concat_imgs_pair = torch.cat([imgs.tensors / 255, ref_img / 255], dim=1)
            flow = self.flownet(concat_imgs_pair)
            warped_feat = self.resample(self.features[0], flow)

            final_feat = warped_feat
        elif infos['frame_category'] == 2: # new key frame 
            self.end_id = min(self.end_id + 1, self.seg_len - 1)
            self.images = deque(maxlen=self.all_frame_interval)
            self.features = deque(maxlen=1)

            feats_cur = self.backbone(imgs.tensors)[0]
            while len(self.images) < self.key_frame_location + 1:
                update_feature(imgs.tensors, feats_cur)
            
            final_feat = feats_cur
        else:
            raise ValueError("Not support frame_category value.")

        final_feat = torch.unsqueeze(final_feat, dim=1)
        proposals, proposal_losses = self.rpn(imgs, final_feat, None)
        if self.roi_heads:
            x, result, detector_losses = self.roi_heads(final_feat, proposals, None)
        else:
            result = proposals

        return result
`
And I use deafult training configs the same as FGFA in your released configs. I am confused is that right? But I cannot get the same result as DFF paper.Some advices? Very thanks

**Comment by Scalsol** [comment]
I have re-implemented DFF and this gives mAP of 74.9%. I think some part of your implementation may be wrong. I will update this project and upload DFF's weight so you will be able to see what mistake you've made.

**Comment by TyroneLi** [comment]
> I have re-implemented DFF and this gives mAP of 74.9%. I think some part of your implementation may be wrong. I will update this project and upload DFF's weight so you will be able to see what mistake you've made.

Thx, cannot wait to see your update, wish that much soon or you could just upate dff code firstly.haha

**Comment by Scalsol** [comment]
DFF has been added. You could refer to the commit history to see my implementation, which is a slightly different from the original implementation. Maybe an instruction for how to implementing new method is needed. I will update it someday.

**Comment by TyroneLi** [comment]
> DFF has been added. You could refer to the commit history to see my implementation, which is a slightly different from the original implementation. Maybe an instruction for how to implementing new method is needed. I will update it someday.

Awesome job! I notice that I missed a sacle factor return from flownet in my original re-implementation.
